### PR TITLE
QAR-202 ArcGISArView behavioural and API changes

### PR DESCRIPTION
**Pull request by LDuncAndroid**
Feature issue: https://devtopia.esri.com/runtime/queen-annes-revenge/issues/202

A number of changes are occurring in this PR:

- Changes to location handling
- Moved starting of ARCore and LocationDataSource to separate functions to allow independent starting/stopping/resetting
- Removed compensation Quaternion as it's no longer required so long as the user sets the pitch of the initial camera to `90.0`

### CODE REVIEW

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt [hunk 1]
```diff
@@ -249,20 +234,26 @@ class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdateListen
...
+    private var useLocationDataSourceOnce: Boolean = true
+
     /**
      * This listener is added to every [LocationDataSource] used when using the [locationDataSource] property to receive
      * location updates.
      *
-     * Upon receiving an updated location, if we don't previously have an [initialLocation] we use that property to
+     * Upon receiving an updated location, if we don't previously have an [originCamera] we use the new location to
      * create a [Camera] and set that as the origin camera of the [cameraController]. On every other location update, if we
      * are not using ARCore we set the current viewpoint camera of [sceneView] to a new Camera created from that location.
```
[the remark this answers is not in the document]
Comment (LDuncAndroid) [reply arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Thank you.

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt [hunk 2]
```diff
@@ -147,36 +142,26 @@ class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdateListen
...
-    private var initialLocation: Point? = null
+    var initialTransformationMatrix: TransformationMatrix = identityMatrix
 
     /**
      * The camera controller used to control the camera that is used in [arcGisSceneView].
      *
      * @since 100.6.0
      */
-    private val cameraController: TransformationMatrixCameraController =
+    private var cameraController: TransformationMatrixCameraController =
```
[the remark this answers is not in the document]
Comment (LDuncAndroid) [reply arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Good point. Left over from when it was public.

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt [hunk 3]
```diff
@@ -271,12 +262,28 @@ class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdateListen
...
-                    sceneView.setViewpointCamera(camera)
+                // Always set the origin camera; then reset ARCore
+                val oldCamera = cameraController.originCamera
+
+                // Create a new camera based on our location and set it on the cameraController.
+                if (originCamera == null) {
+                    val newCamera = Camera(location, 0.0, 90.0, 0.0)
+                    originCamera = newCamera
+                } else {
+                    cameraController.originCamera =
```
[the remark this answers is not in the document]
Comment (LDuncAndroid) [reply arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Thanks for reviewing Mark, sorry, not understanding the comment. Can you elaborate?

[the remark this answers is not in the document]
Comment (LDuncAndroid) [reply arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Ohh, got it 👍

**Comment by LDuncAndroid** [comment]
Thanks for the review @mhdostal . Assigning to @gunt0001 for 2nd review.

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt [hunk 4]
```diff
@@ -147,29 +142,19 @@ class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdateListen
...
-     * transformation for each frame.
-     *
-     * @since 100.6.0
-     */
-    private val compensationQuaternion: Quaternion =
-        Quaternion((sin(45 / (180 / PI)).toFloat()), 0F, 0F, (cos(45 / (180 / PI)).toFloat()))
-
-    /**
-     * Initial location from [locationDataSource].
+     * The camera controller used to control the camera that is used in [arcGisSceneView].
```
Comment (gunt0001) [inline arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Should this line be removed?

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt [hunk 5]
```diff
@@ -249,34 +234,56 @@ class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdateListen
...
-                    val camera = sceneView.currentViewpointCamera.moveTo(location)
-                    sceneView.setViewpointCamera(camera)
+                // Always set the origin camera; then reset ARCore
+                val oldCamera = cameraController.originCamera
+
+                // Create a new camera based on our location and set it on the originCamera.
+                originCamera = if (originCamera == null) {
+                    Camera(location, 0.0, 90.0, 0.0)
+                } else {
+                    Camera(location, oldCamera.heading, oldCamera.pitch, oldCamera.roll)
```
Comment (gunt0001) [inline arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Should this only happen if we are not using ARCore? Based on the doc above...
```
On every other location update, if we are not using ARCore we create a new Camera using properties of the old camera.
```

Or does the doc need to be updated?

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt [hunk 6]
```diff
@@ -469,59 +505,77 @@ class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdateListen
...
+        // Throw exception if Context is not an instance of Activity as it's required for permission request
+        check(context is Activity) { "Context must be an instance of Activity" }
+
+        // ARCore requires camera permissions to operate. If we did not yet obtain runtime
+        // permission on Android M and above, now is a good time to ask the user for it.
+        // when the permission is requested and the user responds to the request from the OS this is executed again
+        // during onResume()
+        if (isUsingARCore == ARCoreUsage.YES && renderVideoFeed && !hasPermission(
+                CAMERA_PERMISSION
+            )
```
Comment (gunt0001) [inline arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Strange line break, can this all go on one line?

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt (hunk 4 above)
Comment (LDuncAndroid) [reply arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Yes, merge artifact.

File: arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt (hunk 6 above)
Comment (LDuncAndroid) [reply arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt]: Yeah, will fix.

**Comment by LDuncAndroid** [comment]
Thanks for the reviews @mhdostal , @gunt0001 . Merging.

**Pull request merged**